core/src/toga/statusicons.py

```diff
@@ -0,0 +1,247 @@
+from __future__ import annotations
+
+import sys
+from collections.abc import Iterator
+from typing import TYPE_CHECKING, Mapping, Sequence
+
+import toga
+from toga.command import Command, CommandSet, Group
+from toga.handlers import wrapped_handler
+from toga.icons import Icon
+from toga.platform import get_platform_factory
+
+if TYPE_CHECKING:
+    from toga.icons import IconContentT
+
+_py_id = id
+
+
+class BaseStatusIcon:
+    def __init__(self, icon: IconContentT | None = None, **kwargs):
+        """An abstract base class for all status icons."""
+        super().__init__(**kwargs)
+        self.factory = get_platform_factory()
+        self._impl = getattr(self.factory, self.__class__.__name__)(interface=self)
+
+        self.icon = icon
+
+    @property
+    def icon(self) -> Icon | None:
+        """The Icon to display in the status bar.
+
+        When setting the icon, you can provide either an :any:`Icon` instance, or a
+        path that will be passed to the ``Icon`` constructor.
+        """
+        return self._icon
+
+    @icon.setter
+    def icon(self, icon_or_name: IconContentT | None):
+        if isinstance(icon_or_name, Icon) or icon_or_name is None:
+            self._icon = icon_or_name
+        else:
+            self._icon = Icon(icon_or_name)
+
+        self._impl.set_icon(self._icon)
+
+
+class StatusIcon(BaseStatusIcon):
+    def __init__(
+        self,
+        id: str | None = None,
+        icon: IconContentT | None = None,
+        text: str | None = None,
+        on_press: toga.widgets.button.OnPressHandler | None = None,
+    ):
+        """
+        An button in a status bar or system tray.
+
+        When pressed, the ``on_press`` handler will be activated.
+
+        :param id: An identifier for the status icon.
+        :param icon: The icon, or icon resource, that will be displayed in the status
+            bar or system tray.
+        :param text: A text label for the status icon. Defaults to the formal name of
+            the app.
+        :param on_press: The handler to invoke when the status icon is pressed.
+        """
+        super().__init__(icon=icon)
+        self.on_press = on_press
+
+        self._id = f"statusicon-{_py_id(self)}" if id is None else id
+        self._text = text if text is not None else toga.App.app.formal_name
+
+    def __repr__(self):
+        return f"<StatusIcon {self.text!r}: {self.id}>"
+
+    @property
+    def id(self) -> str:
+        """A unique identifier for the icon."""
+        return self._id
+
+    @property
+    def text(self) -> str:
+        """A text label for the icon."""
+        return self._text
```

Comment:
I understand it's been done this way because of the duplicate property names in `Group`, but for the purposes of the documentation, the `id` and `text` properties should appear in the base class.